miniapps/solvers/bramble_pasciak.hpp

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2023-2023, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+//
+//          ----------------------------------------------------------
+//               Bramble-Pasciak preconditioning for Darcy problem
+//          ----------------------------------------------------------
+//
+// Main idea is to transform the block system
+//                 Ax = [ M  B^T ] [u] = [f] = b
+//                      [ B   0  ] [p] = [g]
+//     where:
+//        M = \int_\Omega (k u_h) \cdot v_h dx,
+//        B = -\int_\Omega (div_h u_h) q_h dx,
+//        u_h, v_h \in R_h (Raviart-Thomas finite element space),
+//        q_h \in W_h (piecewise discontinuous polynomials),
+// with a block transformation of the form X = A*N - Id
+//                  X = [ M*invQ - Id    0   ]
+//                      [     B*invQ    -Id  ]
+// where N is defined by
+//                  N = [ invQ    0 ]
+//                      [   0     0 ]
+// and Q is constructed such that Q and M-Q are both s.p.d..
+//
+// The solution x is then obtained by solving XAx = Xb with PCG as XA is s.p.d.
+//
+// The codes allows the user to provide such Q, or to construct it from the
+// element matrices M_T. Moreover, the user can provide a block preconditioner
+//                  P = [ M_0    0  ]
+//                      [  0    M_1 ]
+// for the transformed system XA.
+//
+// The code also allows the user to use BPCG, which is a special implementation
+// of the PCG iteration with the particular preconditioner H, defined as
+//                  H = [ M - Q    0  ]
+//                      [  0      M_1 ]
+// BPCG is efficient as it avoids the direct application of invH and X.
```

Comment:
Should the "PCG" and "BPCG" versions give the same results/iterations, just BPCG is the more efficient implementation? Just wanted to clarify what is going on here.

Reply:
In theory. If we precondition the system `XA x = Xb` with `H`, and use PCG, then it should be equivalent to BPCG.